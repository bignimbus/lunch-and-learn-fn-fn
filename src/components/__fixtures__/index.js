export const user = {
  username: 'foo',
};
