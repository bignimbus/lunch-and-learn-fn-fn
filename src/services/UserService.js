class UserService {
  static getUser () {
  }
}

export default UserService;
