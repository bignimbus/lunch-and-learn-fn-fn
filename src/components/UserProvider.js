import React, { Component } from 'react';
import PropTypes from 'prop-types';
import UserService from '../services/UserService';

class UserProvider extends Component {
  static propTypes = {
    WrappedComponent: PropTypes.func.isRequired,
  };

  state = {
    user: {},
  };

  componentDidMount () {
    this.getUser();
  }

  async getUser () {
    const user = await UserService.getUser();
    this.setState({ user });
  }

  render () {
    const { user } = this.state;
    const { WrappedComponent } = this.props;

    return (
      <div>
        {
          user ?
            <WrappedComponent user={user} /> :
            <span>Loading...</span>
        }
      </div>
    );
  }
}

export default UserProvider;
