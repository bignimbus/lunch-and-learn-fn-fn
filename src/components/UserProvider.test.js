import React from 'react';
import renderer from 'react-test-renderer';
import UserProvider from './UserProvider';
import UserService from '../services/UserService';
import { user } from './__fixtures__'

const TestComponent = () => <article id='my-test' />;

const renderComponent = () => renderer.create(
  <UserProvider WrappedComponent={TestComponent} />,
);

describe('UserProvider', () => {
  test('#getUser', async () => {
    UserService.getUser = jest.fn().mockResolvedValue(user);
    const component = renderComponent();
    const instance = component.getInstance();
    await instance.getUser();
    const { state } = instance;
    expect(state.user).toEqual(user);
  });

  describe('#render', () => {
    test('when user has not been fetched', async () => {
      UserService.getUser = jest.fn().mockResolvedValue(null);
      const component = renderComponent();
      await component.getInstance().getUser();
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('when user has been fetched', async () => {
      UserService.getUser = jest.fn().mockResolvedValue(user);
      const component = renderComponent();
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
