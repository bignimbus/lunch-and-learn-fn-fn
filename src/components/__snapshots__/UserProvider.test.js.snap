// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProvider #render when user has been fetched 1`] = `
<div>
  <article
    id="my-test"
  />
</div>
`;

exports[`UserProvider #render when user has not been fetched 1`] = `
<div>
  <span>
    Loading...
  </span>
</div>
`;
